velox/functions/sparksql/specialforms/GetStructField.cpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/specialforms/GetStructField.h"
+#include "velox/expression/ConstantExpr.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+// Returns the value of nested subfield in the input struct.
+// The input must be of row type and nested complex type is allowed.
+// The subfield position is specified by 'ordinal'.
+// If 'ordinal' is negative or greater than the children size of input,
+// exception is thrown.
+class GetStructFieldFunction : public exec::VectorFunction {
+ public:
+  GetStructFieldFunction(int32_t ordinal) : ordinal_(ordinal) {}
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    exec::LocalDecodedVector decoded(context, *args[0], rows);
+    auto rowData = decoded->base()->as<RowVector>();
+    VELOX_USER_CHECK_LT(
```

Comment:
Move the check after
```
VELOX_USER_CHECK_EQ(
      args[0]->type()->kind(),
      TypeKind::ROW,
      "The first argument of get_struct_field should be of row type.");
```
check the row type children size

Reply:
done

Reply:
BTW, it's not ok to move it since when we get rowvector's childrenSize, we should make sure it's decoded. So I move it back

Reply:
The decoded vector type is different with the row type?